### ISSUE TITLE
Gradients are zero when run on GPU (x.cuda())

### ISSUE THREAD

**Issue by void42** [pytorch/pytorch#560]
For example:
```
xx = Variable(torch.randn(1,1), requires_grad=True)
yy = 3*xx
zz = yy**2
zz.backward()
print(xx.grad)
```
yields
```
Variable containing:
-19.2122
[torch.FloatTensor of size 1x1]
```

but

```
xx = Variable(torch.randn(1,1), requires_grad = True).cuda()
yy = 3*xx
zz = yy**2
zz.backward()
print(xx.grad)
```
yields
```
Variable containing:
 0
[torch.cuda.FloatTensor of size 1x1 (GPU 0)]
```

NVIDIA Driver 375.26
CUDA 8.0.44
cuDNN 5.1.3
GTX 1080
Python 3.5.2

**Comment by colesbury** [comment]
You're looking at the gradient of an intermediate computation which are not stored. To fix this, move the `.cuda()` inside the Variable()

```python
xx = Variable(torch.randn(1,1).cuda(), requires_grad = True)
yy = 3*xx
zz = yy**2
zz.backward()
print(xx.grad)
```

What you originally wrote, does the `.cuda()` call **on** the Variable which produces a new Variable. It's equivalent to:


```python
xx = Variable(torch.randn(1,1), requires_grad = True)  # will have .grad stored
xx_cuda = xx.cuda()  # will not have .grad stored
yy = 3*xx_cuda
zz = yy**2
zz.backward()
print(xx.grad)
```

**Comment by void42** [comment]
That makes a lot of sense. Thanks!

**Comment by soumith** [comment]
to read more on getting intermediate stage gradients, you can look at this discussion:
https://discuss.pytorch.org/t/why-cant-i-see-grad-of-an-intermediate-variable/94/10

**Comment by void42** [comment]
I was actually looking at that and was scratching my head. It didn't occur to me that .cuda() was an immutable operation adding a node to the compute graph.
